include/internal/quic_demux.h

```diff
@@ -237,6 +237,41 @@ void ossl_quic_demux_unregister_by_cb(QUIC_DEMUX *demux,
                                       ossl_quic_demux_cb_fn *cb,
                                       void *cb_arg);
 
+/*
+ * Sets a callback for stateless reset processing.
+ *
+ * This callback, if set, is called for every received datagram. info->token and
+ * info->token_len points to a sequence of bytes which are potentially a
+ * stateless reset token. (In current versions of QUIC, token_len is always 16.)
+ * info->peer points to the remote address from which the datagram was received.
+ *
+ * The callback must return 0 if info->token is not a stateless reset token for
+ * the given source IP and 1 if it is. If this function returns 1, the datagram
+ * is not processed as an ordinary packet and not passed to any QRX instance,
+ * and the processing of the datagram stops. The callback is expected to note
+ * that a stateless reset has occurred and perform whatever work is necessary to
+ * handle this case. If the callback returns 0, the datagram is processed
+ * normally and handed off to a QRX instance if possible.
+ *
+ * The info structure and the fields referenced from it are guaranteed to be
+ * valid only for the duration of the callback.
+ *
+ * IMPORTANT: The implementation of this callback must not expose timing
+ * channels for stateless reset tokens.
+ */
+typedef struct quic_demux_reset_info_st {
+    const unsigned char    *token;
+    size_t                  token_len;
+    const BIO_ADDR         *peer;
```

Comment:
Not sure if this is the place, but converting _peer_ to a connection repeatedly isn't good.  I'd have thought QUIC_CONNECTION was a better thing to pass through.  This information is required for non-reset processing regardless.

I think this would require some additional smarts in demux_identify_conn().  Try to determine the connection before checking for reset.

Reply:
This stateless reset callback is called *before* trying to identify a connection by DCID. Stateless reset packets don't have a DCID, so we can't match via that.

Reply:
Doesn't the reset apply to a connection?  We ought to try to figure this out before the callback.
We won't have the CID, but we have the source, port pair which should map to the CID 1-1.
Or am I mistaken?  Server side might not....

Reply:
We need to be able to quickly match the token to the connection on the server so there is really no point to do it differently for the client. There will need to be a global per-SSL_EVENT_CTX mapping of these stateless reset tokens to connections.

Reply:
A UDP port can host multiple connections, both on the server side and the client side. While it might be the case in practice that we use a single connection per port for client-side connections for the foreseeable future it's not required to be the case. In any case, nothing in the demux is designed to demux on network address, so it would be a substantial change to start adding this, and it doesn't seem to make much sense to me.

From the RFC:

> The endpoint identifies a received datagram as a Stateless Reset by comparing the last 16 bytes of the datagram with all stateless reset tokens associated with the remote address on which the datagram was received.

Reply:
> There will need to be a global per-SSL_EVENT_CTX mapping of these stateless reset tokens to connections.

I think it will just be global.  I.e. just one for everything.  Potentially a nasty performance bottleneck.

@hlandau, from that part of the RFC, I don't think we need the network address in the callback.  Just the 16 bytes.